x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```
@@ -0,0 +1,313 @@
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier;
+        if (inferenceId != null) {
+            InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+                TaskType.ANY,
+                inferenceId,
+                List.of(query),
+                Map.of(),
+                InputType.SEARCH
+            );
+
+            inferenceResultsSupplier = new SetOnce<>();
+            queryRewriteContext.registerAsyncAction(
+                (client, listener) -> executeAsyncWithOrigin(
+                    client,
+                    ML_ORIGIN,
+                    InferenceAction.INSTANCE,
+                    inferenceRequest,
+                    listener.delegateFailureAndWrap((l, inferenceResponse) -> {
+                        inferenceResultsSupplier.set(inferenceResponse.getResults());
+                        l.onResponse(null);
+                    })
+                )
+            );
+        } else {
+            // The most likely reason for a null inference ID is an invalid field name, invalid index name(s), or a combination of both.
+            // Set the inference results to an empty list so that query rewriting can continue.
+            // Invalid index names will be handled in TransportSearchAction, which will throw IndexNotFoundException.
+            // Invalid field names will be handled later in the rewrite process.
+            inferenceResultsSupplier = new SetOnce<>(new SparseEmbeddingResults(List.of()));
+        }
+
+        return new SemanticQueryBuilder(this, inferenceResultsSupplier, null);
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        throw new IllegalStateException(NAME + " should have been rewritten to another query type");
+    }
+
+    private static String getInferenceIdForForField(Collection<IndexMetadata> indexMetadataCollection, String fieldName) {
```

@jimczi Per your request, I rewrote this method to fail fast on multiple inference IDs